enterprise/server/remote_execution/copy_on_write/copy_on_write.go

```diff
@@ -1355,3 +1353,93 @@ func checkBounds(opName string, storeSize int64, p []byte, off int64) error {
 	}
 	return nil
 }
+
+// BoundedStack is a concurrency-safe stack with bounded size: when the
+// capacity of the stack is exceeded, the least recently added item ("bottom")
+// of the stack is removed.
+//
+// Its intended usage is to prioritize the most recently added items for eager
+// chunk fetching.
+//
+// It exposes a channel C which streams a value whenever a value is pushed on
+// the stack. Example:
+//
+//	for range stack.C {
+//		val, ok := stack.Pop()
+//		if !ok {
+//			// A receive from C normally guarantees that a value is ready,
+//			// but it's best to always check `ok` when popping from the stack.
+//			continue
+//		}
+//		// Do something with val...
+//	}
+type BoundedStack[T any] struct {
+	C chan struct{}
+
+	mu     sync.Mutex
+	buffer []T
+	top    int
+	size   int
+}
+
+func NewBoundedStack[T any](capacity int) (*BoundedStack[T], error) {
+	if capacity <= 0 {
+		return nil, status.InvalidArgumentError("bounded stack capacity must be > 0")
+	}
+	return &BoundedStack[T]{
+		C:      make(chan struct{}, 1),
+		buffer: make([]T, capacity),
+	}, nil
+}
+
+// Push adds an item to the top of the stack, evicting the bottom item of the
+// stack if the stack is full.
+func (s *BoundedStack[T]) Push(item T) {
+	s.push(item)
+	// Notify the channel that a new item was pushed.
+	select {
+	case s.C <- struct{}{}:
+	default:
+	}
+}
+
+func (s *BoundedStack[T]) push(item T) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.top = (s.top + 1) % len(s.buffer)
+	s.buffer[s.top] = item
+	s.size = min(s.size+1, len(s.buffer))
+}
+
+// Pop returns the top element of the stack. The second return value indicates
+// whether the returned item is valid; it will be false if and only if the stack
+// was empty.
+func (s *BoundedStack[T]) Pop() (item T, ok bool) {
+	val, newSize, ok := s.pop()
+	if newSize > 0 {
+		// Re-saturate the channel if the stack still has items after popping,
+		// so that another goroutine can pop the next element.
+		select {
+		case s.C <- struct{}{}:
```

Comment:
I'm kind of confused about what this is supposed to do. Are we not double adding to the channel for each element in the stack? We already added to the channel in Push()

Reply:
(See other comment)

Reply:
Actually NVM, `struct{}` consumes 0 bytes of memory so this is just straight up better.